prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +24,108 @@ def floatify(value, default=float("nan")):
         return default
 
 
-def map_generator_to_sub_by_location(generator, substation_groupby):
-    """Determine a likely substation for a generator to be connected to. Priority order
-    of mapping is: 1) if location is available and one or more substations exist in that
-    ZIP code, map by location to closest substation within that ZIP code, 2) if location
-    is available but no substations exist in that ZIP code, map to the closest
-    substation within neighboring ZIP codes, 3) if only ZIP code is available
-    (no location), and one or more substations exist, map to an arbitrarily chosen
-    substation within that ZIP code, 4) if only ZIP code is available (no location)
-    but no substations exist in that ZIP code, return NA.
-
-    :param pandas.Series generator: one generating unit from data frame.
-    :param pandas.GroupBy substation_groupby: data frame of substations, grouped by
-        (interconnect, ZIP).
-    :return: (*int/pd.NA*) -- substation ID if the generator can be mapped successfully
-        to a substation, else pd.NA.
+def map_generators_to_sub_by_location(generators, substations, inplace=True):
+    """Determine the closest substation to each generator. For generators without
+    latitude and longitude, an attempt will be made to match via ZIP code, and failing
+    that a pandas.NA value will be returned.
+
+    :param pandas.DataFrame generators: generator data. Required columns:
+        'interconnect', 'lat', 'lon', 'state', 'ZIP'.
+    :param pandas.DataFrame substations: substation data. Required columns:
+        'interconnect', 'lat', 'lon', 'STATE', 'ZIP'.
+    :param bool inplace: whether to modify the generator table inplace with a new
+        'sub_id' column or to return a new one. If ``inplace`` is `True`, entries in
+        `generators` which have non-sensical combinations of 'state' and 'interconnect'
```

Comment:
To clarify the logic in the PR description: we will always trust the `state` tag of the generator. For boundary states, generators connect to other interconnections (different from the one specified in the `interconnect` column) is allowed, otherwise, the corresponding interconnect of the generator will be fixed, right?

Reply:
For 'boundary' states, we trust that the generator may actually be in the listed state, but that it may be connected to a substation in the matching interconnection but within another state. E.g. a generator in Oklahoma may actually be connected to the bulk power grid via an ERCOT substation in Texas. So from the grid's perspective, we trust the `interconnect` column and lat/lon more than the `state` column (since the grid only really cares where the power gets injected), although we do use the `state` column to determine when to apply this logic.

For non-boundary states, we trust the `state` column and the lat/lon more than the `interconnect` column, which we ignore.

Reply:
My question is more on what will be potentially changed with `inplace` set, i.e. for boundary states, we will trust the `interconnect`, thus both `state` and `interconnect` won't be changed, whereas for non-boundary states, we will trust the `state` and `interconnect` will be changed if does seem to be correct, right?

Reply:
Right. If the generator is in Virginia, then we don't trust the `interconnect` column that says `Western` (which comes from EIA Form 860 when it says it's connected to CISO), and we change it to 'Eastern' instead.